### PR TITLE
Implemented Visual Deep Neural Network Builder

### PR DESCRIPTION
**Pull request by Lakini**
GSoC Project 2016.
Proposal 03 : [ML] Visual Deep Neural Network Builder for WSO2 Machine Learner.

### CODE REVIEW

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java [hunk 1]
```diff
@@ -0,0 +1,511 @@
+/*
+ * Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Year should be 2016. Please fix on all places

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+
+    OptimizationAlgorithm mapOptimizationAlgorithm(String optimizationAlgorithms){
+
+        OptimizationAlgorithm optimizationAlgo = null;
+
+        switch (optimizationAlgorithms){
+            case "Line_Gradient_Descent":
+                optimizationAlgo = OptimizationAlgorithm.LINE_GRADIENT_DESCENT;
+                break;
+
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Remove unnecessary newlines

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java [hunk 3]
```diff
@@ -614,4 +620,78 @@ public Response getUserRecommendations(@PathParam("modelId") long modelId,
...
+            statistics = net.createFeedForwardNetwork(seed, learningRate, bachSize, nepoches, iterations, optimizationAlgorithms, updater, momentum, pretrain, backprop, noHiddenLayers, inputLayerNodes, datasetId, versionId, analysisId, hiddenLayerList, outputLayerList);
+
+            ObjectMapper objectMapper = new ObjectMapper();
+            Object statJson = objectMapper.readValue(objectMapper.writeValueAsString(statistics), Object.class);
+            logger.info("API Response " + statJson.toString());
+
+            return Response.ok(statJson).build();
+
+        } catch (Exception e) {
+            String msg = MLUtils.getErrorMsg("Error occurred in the server side!!!", e);
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java]: Use a proper error message. Indicate what the error is.

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java [hunk 4]
```diff
@@ -614,4 +620,78 @@ public Response getUserRecommendations(@PathParam("modelId") long modelId,
...
+            //Call createFeedForwardNetwork method
+            statistics = net.createFeedForwardNetwork(seed, learningRate, bachSize, nepoches, iterations, optimizationAlgorithms, updater, momentum, pretrain, backprop, noHiddenLayers, inputLayerNodes, datasetId, versionId, analysisId, hiddenLayerList, outputLayerList);
+
+            ObjectMapper objectMapper = new ObjectMapper();
+            Object statJson = objectMapper.readValue(objectMapper.writeValueAsString(statistics), Object.class);
+            logger.info("API Response " + statJson.toString());
+
+            return Response.ok(statJson).build();
+
+        } catch (Exception e) {
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java]: Catch specific exceptions.

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+                .optimizationAlgo(mapOptimizationAlgorithm(optimizationAlgorithms))
+                .learningRate(learningRate)
+                .updater(mapUpdater(updater))
+                .momentum(momentum)
+                .list(noHiddenLayers+1);
+
+        //Add Hidden Layers to the network with unique settings
+        for(int i = 0;i< noHiddenLayers;i++){
+            int nInput = 0;
+            if(i == 0)
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: use curly braces after if/else, even if its a single line

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+        switch (optimizationAlgorithms){
+            case "Line_Gradient_Descent":
+                optimizationAlgo = OptimizationAlgorithm.LINE_GRADIENT_DESCENT;
+                break;
+
+            case "Conjugate_Gradient":
+                optimizationAlgo = OptimizationAlgorithm.CONJUGATE_GRADIENT;
+                break;
+
+            case "Hessian_Free":
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Use enums

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+
+        Updater updaterAlgo = null;
+
+        switch (updater) {
+
+            case "sgd":
+                updaterAlgo = Updater.SGD;
+                break;
+
+            case "adam":
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: use enums

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+     * method to map user selected Loss Function Algorithm to LossFunction object.
+     * @param lossFunction
+     * @return an LossFunction object .
+     */
+    LossFunction mapLossFunction(String lossFunction){
+
+        LossFunction lossfunctionAlgo = null;
+
+        switch (lossFunction){
+            case "mse":
```
Comment (SupunS) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: use enums

File: pom.xml [hunk 9]
```diff
@@ -569,6 +569,49 @@
...
+                <artifactId>deeplearning4j-core</artifactId>
+                <version>${dl4j.core.version}</version>
+            </dependency>
+
+            <!-- ND4J-->
+            <!-- https://mvnrepository.com/artifact/org.nd4j/nd4j-api -->
+            <dependency>
+                <groupId>org.nd4j</groupId>
+                <artifactId>nd4j-api</artifactId>
+                <version>0.4.0</version>
```
Comment (SupunS) [inline pom.xml]: Define the version under the <properties> tag. Refer other dependencies for example.

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java (hunk 2 above)
Comment (Lakini) [inline components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Fixed in https://github.com/wso2/carbon-ml/pull/233/commits/275de09f616359250d4252a996d9ef255bc68def

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java (hunk 3 above)
Comment (Lakini) [reply components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java]: fixed in commit 60f3de25ba419dc20e044130ba2dc7e324b06570

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java (hunk 4 above)
Comment (Lakini) [reply components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/ModelApiV20.java]: Fixed in 60f3de25ba419dc20e044130ba2dc7e324b06570

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java (hunk 5 above)
Comment (Lakini) [reply components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Fixed in 60f3de25ba419dc20e044130ba2dc7e324b06570

File: pom.xml (hunk 9 above)
Comment (Lakini) [reply pom.xml]: Fixed in 60f3de25ba419dc20e044130ba2dc7e324b06570

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java (hunk 6 above)
Comment (Lakini) [reply components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Fixed in 3a23d9ab7a4e51a95207398e8fdeacd0198a1642

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java (hunk 7 above)
Comment (Lakini) [reply components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Fixed in 3a23d9ab7a4e51a95207398e8fdeacd0198a1642

File: components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java (hunk 8 above)
Comment (Lakini) [reply components/ml/org.wso2.carbon.ml.rest.api/src/main/java/org/wso2/carbon/ml/rest/api/neuralNetworks/FeedForwardNetwork.java]: Fixed in 3a23d9ab7a4e51a95207398e8fdeacd0198a1642

**Comment by nirmal070125** [comment]
Thank you for your contribution. We are not in a position to merge this as we have stopped feature addition to ML.

**Pull request closed without merging**